File: src/algorithms/uniform.js
import classifyPoint from "robust-point-in-polygon";

import { ACRE, SQUARE_SIDE, BOUNDARY_THRESHOLD } from "./constants";
import {
  boundingBox,
  distanceToLine,
  lngLatPoints,
  utmPoints,
  utmString,
} from "./utils";

const generateUniformGrid = (bbox) => {
  const [xmin, ymin, xmax, ymax] = bbox;
  let i = 1;
  let j;
  const grid = [];
  while (xmin + (i * SQUARE_SIDE[ACRE]) / 2 < xmax) {
    j = 1;
    while (ymin + (j * SQUARE_SIDE[ACRE]) / 2 < ymax) {
      grid.push([
        xmin + (i * SQUARE_SIDE[ACRE]) / 2,
        ymin + (j * SQUARE_SIDE[ACRE]) / 2,
      ]);
      j++;
    }
    i++;
  }
  return grid;
};

export function uniformGrid(polygon) {
  const projectionString = utmString(polygon[0]);
  const utm = utmPoints(polygon, projectionString);
  const lines = [];
  for (let i = 0; i < utm.length - 1; i++) {
    lines.push([utm[i][0], utm[i][1], utm[i + 1][0], utm[i + 1][1]]);
  }
  let grid = generateUniformGrid(boundingBox(utm));
  // console.log(grid.length);
  grid = grid.filter((point) => classifyPoint(utm, point) === -1);
  // console.log(grid.length);
  grid = grid.filter((point) => {
    const minDistance = Math.min(
      ...lines.map((line) => distanceToLine(point, line))
    );
    if (minDistance < BOUNDARY_THRESHOLD) {
      return false;
    }
    return true;
  });
  // console.log(grid.length);
  const out = lngLatPoints(grid, projectionString);
  return out;
}

File: api/sampling/uniform.py
import numpy as np
from shapely.geometry import Polygon, Point

SQUARE_SIDE = {
    "1": 63.615,
    "2.5": 100.584,
    "5": 142.247,
}


def in_polygon_filter(grid: np.ndarray, polygon: np.ndarray) -> np.ndarray:
    """Filter out grid points that are not in the polygon."""
    shapely_polygon = Polygon(polygon)
    keepers = []
    for point in grid:
        if shapely_polygon.contains(Point(point)):
            keepers.append(point)
    return np.array(keepers)


def boundary_distance_filter(
    grid: np.ndarray, polygon: np.ndarray, acre: str
) -> np.ndarray:
    """Filter out grid points that are within <threshold> of the polygon boundary."""
    shapely_polygon = Polygon(polygon)
    boundary_threshold = SQUARE_SIDE[acre] / 2
    keepers = []
    for point in grid:
        if shapely_polygon.exterior.distance(Point(point)) > boundary_threshold:
            keepers.append(point)
    return np.array(keepers)


def uniform_sample(polygon: np.ndarray, acre: str = "1") -> np.ndarray:
    """Generate a uniform grid that covers an entire polygon."""
    x_min = polygon[:, 0].min()
    x_max = polygon[:, 0].max()
    y_min = polygon[:, 1].min()
    y_max = polygon[:, 1].max()

    i = 1
    grid = []
    while x_min + i * SQUARE_SIDE[acre] / 2 < x_max:
        j = 1
        while y_min + j * SQUARE_SIDE[acre] / 2 < y_max:
            grid.append(
                [
                    x_min + i * SQUARE_SIDE[acre] / 2,
                    y_min + j * SQUARE_SIDE[acre] / 2,
                ]
            )
            j += 1
        i += 1
    grid = np.array(grid)
    grid = in_polygon_filter(grid, polygon)
    grid = boundary_distance_filter(grid, polygon, acre)
    return grid
